include/deal.II/simplex/barycentric_polynomials.h

```diff
@@ -0,0 +1,639 @@
+                           const Point<dim> & p) const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::compute_4th_derivative()
+     */
+    Tensor<4, dim>
+    compute_4th_derivative(const unsigned int i,
+                           const Point<dim> & p) const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::compute_grad()
+     */
+    Tensor<1, dim>
+    compute_grad(const unsigned int i, const Point<dim> &p) const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::compute_grad_grad()
+     */
+    Tensor<2, dim>
+    compute_grad_grad(const unsigned int i, const Point<dim> &p) const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::memory_consumption().
+     */
+    virtual std::size_t
+    memory_consumption() const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::name()
+     */
+    std::string
+    name() const override;
+
+    /**
+     * @copydoc ScalarPolynomialsBase::clone()
+     */
+    virtual std::unique_ptr<ScalarPolynomialsBase<dim>>
+    clone() const override;
+
+  protected:
+    std::vector<BarycentricPolynomial<dim>> polys;
+
+    Table<2, BarycentricPolynomial<dim>> poly_grads;
+
+    Table<3, BarycentricPolynomial<dim>> poly_hessians;
+
+    Table<4, BarycentricPolynomial<dim>> poly_third_derivatives;
+
+    Table<5, BarycentricPolynomial<dim>> poly_fourth_derivatives;
+  };
+} // namespace Simplex
+
+// free template functions (in Simplex, not dealii::Simplex)
+
+/**
+ * Multiply a Simplex::BarycentricPolynomial by a constant.
+ */
+template <int dim, typename Number1, typename Number2>
+Simplex::BarycentricPolynomial<dim, Number1>
+operator*(const Number2 &                                     a,
+          const Simplex::BarycentricPolynomial<dim, Number1> &bp)
+{
+  return bp * Number1(a);
+}
+
+/**
+ * Add a constant to a Simplex::BarycentricPolynomial.
+ */
+template <int dim, typename Number1, typename Number2>
+Simplex::BarycentricPolynomial<dim, Number1>
+operator+(const Number2 &                                     a,
+          const Simplex::BarycentricPolynomial<dim, Number1> &bp)
+{
+  return bp + Number1(a);
+}
+
+/**
+ * Subtract a Simplex::BarycentricPolynomial from a constant.
+ */
+template <int dim, typename Number1, typename Number2>
+Simplex::BarycentricPolynomial<dim, Number1>
+operator-(const Number2 &                                     a,
+          const Simplex::BarycentricPolynomial<dim, Number1> &bp)
+{
+  return bp - Number1(a);
+}
+
+/**
+ * Write a Simplex::BarycentricPolynomial to the provided output stream.
+ */
+template <int dim, typename Number>
+std::ostream &
+operator<<(std::ostream &                                     out,
+           const Simplex::BarycentricPolynomial<dim, Number> &bp)
+{
+  bp.print(out);
+  return out;
+}
+
+// Template function definitions
+
+namespace internal
+{
+  template <int N>
+  TableIndices<N>
+  index_to_indices(const std::size_t &index, const TableIndices<N> &extent)
```

Comment:
What does this function do?

Reply:
It converts an integral index into something a `Table` can use to look up an entry. I'll document it.